lib/Dialect/TTNN/Analysis/L1InterleavedPolicy.cpp

```diff
@@ -24,132 +28,323 @@ uint64_t getOpOutputLayoutUsage(
-        currentOp = scheduleableOps[0];
+      // Pick optimal L1Precedence among all legal L1Precedence.
+      // The one that requires the least amount of L1 cache overall is
+      // considered optimal.
+      //
+      if (isLegal && intermediateRequiredL1Usage < minRequiredL1Usage) {
+        isMaskLegal = true;
+        minRequiredL1Usage = intermediateRequiredL1Usage;
+        optimalL1Precedence = L1Precedence;
       }
+    } while (std::next_permutation(L1Precedence.begin(), L1Precedence.end()));
+
+    if (isMaskLegal && optimalL1Usage < currentL1Usage &&
+        currentL1Usage <= getAvailableL1CacheSize()) {
 
-      // Schedule currentOp.
+      // Append the legal L1Precedence to the currentPrecedence and therefore
+      // create a complete precedence for the baseOp and currentMask.
       //
-      scheduler.scheduleOp(currentOp);
+      currentPrecedence.insert(currentPrecedence.end(),
+                               optimalL1Precedence.begin(),
+                               optimalL1Precedence.end());
 
-      // Skip starting sharding chain if currentOp is a memory management op.
+      // Update the optimal configuration.
       //
-      if (l1ChainConfigs->back().isEmpty() && isa<ToLayoutOp>(currentOp)) {
-        currentOp = nullptr;
-        continue;
-      }
+      optimalL1Usage = currentL1Usage;
+      optimalLayouts = std::move(currentLayouts);
+      optimalPrecedence = std::move(currentPrecedence);
+    }
+  }
+
+  // Create the optimal config.
+  //
+  Config optimalConfig;
+  optimalConfig.layouts = std::move(optimalLayouts);
+  optimalConfig.precedence = std::move(optimalPrecedence);
 
-      if (scheduler.hasUnscheduledOps()) {
-        scheduleableOps = scheduler.getScheduleableOps();
+  return optimalConfig;
+}
 
-        // Check if currentOp has a valid successor.
+void L1InterleavedPolicy::run() {
+  for (Operation &funcOp : rootOp->getRegion(0).getOps()) {
+    func::FuncOp func = dyn_cast<func::FuncOp>(funcOp);
+    DeviceAttr deviceAttr = getCurrentScopeDevice(func);
+
+    // Start the policy.
+    //
+    llvm::DenseMap<Operation *, OpMemSpec> OpMemSpecMap;
+    mlir::tt::scheduler::Scheduler scheduler(&func);
+    llvm::SmallVector<Operation *> scheduleableOps;
+
+    while (scheduler.hasUnscheduledOps()) {
+      scheduleableOps = scheduler.getScheduleableOps();
+
+      for (Operation *op : scheduleableOps) {
+        // Schedule the op.
         //
-        Operation *nextOp = nullptr;
-        for (auto *op : scheduleableOps) {
-          for (auto operand : op->getOperands()) {
-            if (operand.getDefiningOp() == currentOp) {
-              nextOp = op;
-              break;
-            }
+        scheduler.scheduleOp(op);
+
+        // Find optimal configuration for the op.
+        //
+        llvm::DenseMap<Operation *, L1Usage> opsL1Usage;
+        llvm::SmallVector<Operation *> opsPrecedence;
+
+        // Generate optimal configuration for the current op based on the
```

Comment:
What is the case when this is false?

Reply:
There are two ways that the op is not analysable: 1) Op does not produce any output (e.g. `ttnn.get_device()`) and 2) all op's legal layouts are filtered out in `MemoryLayoutAnalysis` (this is true for the last `ToLayout` op that converts tensor back to `System` memory and therefore all its only legal layout will be filtered out since it not DRAM nor L1 buffer type).